crates/sui-transaction-builder/src/object.rs

```diff
@@ -0,0 +1,203 @@
+        }
+    }
+
+    // Add partial information
+
+    // Kind
+    pub fn as_owned(self) -> Self {
+        Self {
+            id: self.id,
+            kind: Some(Kind::ImmOrOwned),
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
+    }
+    // Redundant, but who ever liked saying "imm_or_owned"?
+    pub fn as_immutable(self) -> Self {
+        Self {
+            id: self.id,
+            kind: Some(Kind::ImmOrOwned),
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: Some(false),
+        }
+    }
+
+    pub fn as_receiving(self) -> Self {
+        Self {
+            id: self.id,
+            kind: Some(Kind::Receiving),
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
+    }
+
+    pub fn as_shared(self) -> Self {
+        Self {
+            id: self.id,
+            kind: Some(Kind::Shared),
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
+    }
+
+    // ObjectRef fields
+    pub fn versioned_at(self, version: u64) -> Self {
+        Self {
+            id: self.id,
+            kind: self.kind,
+            version: Some(version),
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
+    }
+
+    pub fn with_digest(self, digest: ObjectDigest) -> Self {
+        Self {
+            id: self.id,
+            kind: self.kind,
+            version: self.version,
+            digest: Some(digest),
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
+    }
+
+    // Shared fields
+
+    // Initial shared version
+    pub fn shared_at(self, i: u64) -> Self {
+        Self {
+            id: self.id,
+            kind: self.kind,
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: Some(i),
+            mutable: self.mutable,
+        }
+    }
+
+    // Shared value mutability
+    pub fn by_val(self) -> Self {
+        self
+    }
+    pub fn by_ref(&self) -> &Self {
+        self
+    }
+    pub fn by_mut(&mut self) -> &mut Self {
+        self
+    }
+}
+
+impl TryInto<ObjectReference> for &Object {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> Result<ObjectReference, Self::Error> {
+        Ok(ObjectReference::new(
+            self.id,
+            self.version.ok_or_else(|| anyhow!("version not set"))?,
+            self.digest.ok_or_else(|| anyhow!("digest not set"))?,
+        ))
+    }
+}
```

Comment:
Generally `From` impls are preferable to `Into` impls [because `From` implies `Into` but not the other way around](https://doc.rust-lang.org/rust-by-example/conversion/from_into.html#from-and-into-are-interchangable).

And generally, you should avoid conversion operators that take references -- they suggest that there isn't really a conversion going on.

Reply:
Yeah, don't know why I did not do that. 🙃